File: pw_rpc/py/pw_rpc/packets.py
# Copyright 2020 The Pigweed Authors
#
# Licensed under the Apache License, Version 2.0 (the "License"); you may not
# use this file except in compliance with the License. You may obtain a copy of
# the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
# License for the specific language governing permissions and limitations under
# the License.
"""Functions for working with pw_rpc packets."""

import dataclasses
from typing import Optional

from google.protobuf import message
from pw_status import Status

from pw_rpc.internal import packet_pb2


def decode(data: bytes) -> packet_pb2.RpcPacket:
    packet = packet_pb2.RpcPacket()
    packet.MergeFromString(data)
    return packet


def decode_payload(packet, payload_type):
    payload = payload_type()
    payload.MergeFromString(packet.payload)
    return payload


@dataclasses.dataclass(eq=True, frozen=True)
class RpcIds:
    """Integer IDs that uniquely identify a remote procedure call."""

    channel_id: int
    service_id: int
    method_id: int
    call_id: int


def encode_request(rpc: RpcIds, request: Optional[message.Message]) -> bytes:
    payload = request.SerializeToString() if request is not None else bytes()

    return packet_pb2.RpcPacket(
        type=packet_pb2.PacketType.REQUEST,
        channel_id=rpc.channel_id,
        service_id=rpc.service_id,
        method_id=rpc.method_id,
        call_id=rpc.call_id,
        payload=payload,
    ).SerializeToString()


def encode_response(rpc: RpcIds, response: message.Message) -> bytes:
    return packet_pb2.RpcPacket(
        type=packet_pb2.PacketType.RESPONSE,
        channel_id=rpc.channel_id,
        service_id=rpc.service_id,
        method_id=rpc.method_id,
        call_id=rpc.call_id,
        payload=response.SerializeToString(),
    ).SerializeToString()


def encode_client_stream(rpc: RpcIds, request: message.Message) -> bytes:
    return packet_pb2.RpcPacket(
        type=packet_pb2.PacketType.CLIENT_STREAM,
        channel_id=rpc.channel_id,
        service_id=rpc.service_id,
        method_id=rpc.method_id,
        call_id=rpc.call_id,
        payload=request.SerializeToString(),
    ).SerializeToString()


def encode_client_error(packet: packet_pb2.RpcPacket, status: Status) -> bytes:
    return packet_pb2.RpcPacket(
        type=packet_pb2.PacketType.CLIENT_ERROR,
        channel_id=packet.channel_id,
        service_id=packet.service_id,
        method_id=packet.method_id,
        call_id=packet.call_id,
        status=status.value,
    ).SerializeToString()


def encode_cancel(rpc: RpcIds) -> bytes:
    return packet_pb2.RpcPacket(
        type=packet_pb2.PacketType.CLIENT_ERROR,
        status=Status.CANCELLED.value,
        channel_id=rpc.channel_id,
        service_id=rpc.service_id,
        method_id=rpc.method_id,
        call_id=rpc.call_id,
    ).SerializeToString()


def encode_client_stream_end(rpc: RpcIds) -> bytes:
    return packet_pb2.RpcPacket(
        type=packet_pb2.PacketType.CLIENT_REQUEST_COMPLETION,
        channel_id=rpc.channel_id,
        service_id=rpc.service_id,
        method_id=rpc.method_id,
        call_id=rpc.call_id,
    ).SerializeToString()


def for_server(packet: packet_pb2.RpcPacket) -> bool:
    return packet.type % 2 == 0

File: pw_rpc/ts/packets.ts
// Copyright 2022 The Pigweed Authors
//
// Licensed under the Apache License, Version 2.0 (the "License"); you may not
// use this file except in compliance with the License. You may obtain a copy of
// the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
// License for the specific language governing permissions and limitations under
// the License.

/** Functions for working with pw_rpc packets. */

import { Message } from 'google-protobuf';
import {
  RpcPacket,
  PacketType,
} from 'pigweedjs/protos/pw_rpc/internal/packet_pb';
import { Status } from 'pigweedjs/pw_status';
import { MessageSerializer } from './descriptors';

// Channel, Service, Method, CallId
type idSet = [number, number, number, number];

export function decode(data: Uint8Array): RpcPacket {
  return RpcPacket.deserializeBinary(data);
}

function serializeMessage(
  message?: Message,
  serializers?: MessageSerializer,
): Uint8Array {
  let payload: Uint8Array;
  if (typeof message !== 'undefined') {
    if (serializers) {
      payload = serializers.serialize(message);
    } else {
      payload = (message as any)['serializeBinary']();
    }
  } else {
    payload = new Uint8Array(0);
  }
  return payload;
}

export function decodePayload(
  payload: Uint8Array,
  payloadType: any,
  serializer?: MessageSerializer,
): any {
  if (serializer) {
    return serializer.deserialize(payload);
  }
  return payloadType['deserializeBinary'](payload);
}

export function forServer(packet: RpcPacket): boolean {
  return packet.getType() % 2 == 0;
}

export function encodeClientError(
  packet: RpcPacket,
  status: Status,
): Uint8Array {
  const errorPacket = new RpcPacket();
  errorPacket.setType(PacketType.CLIENT_ERROR);
  errorPacket.setChannelId(packet.getChannelId());
  errorPacket.setMethodId(packet.getMethodId());
  errorPacket.setServiceId(packet.getServiceId());
  errorPacket.setCallId(packet.getCallId());
  errorPacket.setStatus(status);
  return errorPacket.serializeBinary();
}

export function encodeClientStream(
  ids: idSet,
  message: Message,
  customSerializer?: MessageSerializer,
): Uint8Array {
  const streamPacket = new RpcPacket();
  streamPacket.setType(PacketType.CLIENT_STREAM);
  streamPacket.setChannelId(ids[0]);
  streamPacket.setServiceId(ids[1]);
  streamPacket.setMethodId(ids[2]);
  streamPacket.setCallId(ids[3]);
  const msgSerialized = serializeMessage(message, customSerializer);
  streamPacket.setPayload(msgSerialized);
  return streamPacket.serializeBinary();
}

export function encodeClientStreamEnd(ids: idSet): Uint8Array {
  const streamEnd = new RpcPacket();
  streamEnd.setType(PacketType.CLIENT_REQUEST_COMPLETION);
  streamEnd.setChannelId(ids[0]);
  streamEnd.setServiceId(ids[1]);
  streamEnd.setMethodId(ids[2]);
  streamEnd.setCallId(ids[3]);
  return streamEnd.serializeBinary();
}

export function encodeRequest(
  ids: idSet,
  request?: Message,
  customSerializer?: MessageSerializer,
): Uint8Array {
  const payload = serializeMessage(request, customSerializer);
  const packet = new RpcPacket();
  packet.setType(PacketType.REQUEST);
  packet.setChannelId(ids[0]);
  packet.setServiceId(ids[1]);
  packet.setMethodId(ids[2]);
  packet.setCallId(ids[3]);
  packet.setPayload(payload);
  return packet.serializeBinary();
}

export function encodeResponse(ids: idSet, response: Message): Uint8Array {
  const packet = new RpcPacket();
  packet.setType(PacketType.RESPONSE);
  packet.setChannelId(ids[0]);
  packet.setServiceId(ids[1]);
  packet.setMethodId(ids[2]);
  packet.setCallId(ids[3]);
  const msgSerialized = (response as any)['serializeBinary']();
  packet.setPayload(msgSerialized);
  return packet.serializeBinary();
}

export function encodeCancel(ids: idSet): Uint8Array {
  const packet = new RpcPacket();
  packet.setType(PacketType.CLIENT_ERROR);
  packet.setStatus(Status.CANCELLED);
  packet.setChannelId(ids[0]);
  packet.setServiceId(ids[1]);
  packet.setMethodId(ids[2]);
  packet.setCallId(ids[3]);
  return packet.serializeBinary();
}
